core/src/banking_stage/immutable_deserialized_packet.rs

```diff
@@ -116,22 +119,37 @@ impl ImmutableDeserializedPacket {
         votes_only: bool,
         address_loader: impl AddressLoader,
         reserved_account_keys: &HashSet<Pubkey>,
-    ) -> Option<SanitizedTransaction> {
+    ) -> Option<RuntimeTransaction<SanitizedTransaction>> {
         if votes_only && !self.is_simple_vote() {
             return None;
         }
-        let tx = SanitizedTransaction::try_new(
-            self.transaction().clone(),
-            *self.message_hash(),
-            self.is_simple_vote(),
-            address_loader,
-            reserved_account_keys,
+
+        let tx = RuntimeTransaction::<SanitizedVersionedTransaction>::try_from(
+            self.transaction.clone(),
+            Some(self.message_hash),
+            Some(self.is_simple_vote),
         )
+        .and_then(|tx| {
+            RuntimeTransaction::<SanitizedTransaction>::try_from(
+                tx,
+                address_loader,
+                reserved_account_keys,
+            )
+        })
```

Comment:
So it has two steps here: creating the "static" version of `RuntimeTransaction::<SanitizedVersionedTransaction>`, then creating the "dynamic" version of `RuntimeTransaction<SanitizedTransaction>`; this illustrates the ideal use of `RuntimeTransaction`, which is: the "static" version can advance to "dynamic" version; and "dynamic" version is the superset of "static" version. 

How do you think to create "static" version in `ImmutableDeserializedPacket::new()`, then consume "static" by advance it into "dynamic" when needed? I know you are expecting IDP to be retired soon, but I think it's worth to update IDP to properly use `RuntimeTransaction` to shake out its usages, but also save one `clone` of `SanitizedVersionedTransaction` at ln 128.

I'm thinking something like this: 
1. change `ImmutableDeserializePacket` def to:
```
enum RuntimeTransactionVersion {
    Static (RuntimeTransaction::<SanitizedVersionedTransaction>),
    Dynamic (RuntimeTransaction::<SanitizedTransaction>),
}

pub struct ImmutableDeserializedPacket {
    original_packet: Packet,
    runtime_transaction: RuntimeTransactionVersion,
    // ^^ it caches the rest of stuff IDP used to have
}
```
2. change `new()` to: 
```
    pub fn new(packet: Packet) -> Result<Self, DeserializedPacketError> {
        let versioned_transaction: VersionedTransaction = packet.deserialize_slice(..)?;
        let sanitized_transaction = SanitizedVersionedTransaction::try_from(versioned_transaction)?;

        Ok(Self {
            original_packet: packet,
            runtime_transaction: Static(RuntimeTransaction::<SanitizedVersionedTransaction>::try_from(
                sanitized_transaction, None, None)),
        })
    }
```
3. So the `build_sanitized_transaction()` would be renamed to: 
```
    pub fn get_sanitized_transaction(
        &mut self,
        votes_only: bool,
        address_loader: impl AddressLoader,
        reserved_account_keys: &HashSet<Pubkey>,
    ) -> Result<()> {
        if votes_only && !self.is_simple_vote() {
            return None;
        }

        if self.runtime_transaction.is_static() {
            // consume static, advance to dynamic
            self.runtime_transaction = Dynamic( 
                RuntimeTransaction::<SanitizedTransaction>::try_from(
                    self.runtime_transaction, address_loader, reserved_account_keys)
            )?;
        }

        Ok(())
    }
```

(sorry for pseudo code may not be practical)

Reply:
could do something like that. enum should be relatively light overhead. i'll test it out tomorrow.

Reply:
Thinking through it this morning. I'm not sure it's worth the added complexity.

If we stick to not implementing `Clone` for `RuntimeTransaction` (could practice imo), then to process the batch we need to take owership of the `RuntimeTransaction`; we'd then need to put the `RuntimeTransaction` back into the `ImmutableDeserializedPacket` on retry.
This adds a ton of complexity and possible bugs.

I think also transitionining from static->dynamic runtime also breaks the naming of `Immutable`.

Reply:
one option is to only implement `Clone` for `RuntimeTransaction<SanitizedTransaction>`; but that feels a bit dirty.

Reply:
Chatted offline, we believed IDP will be refactored out soon, so we'll leave it as is for now.

Reply:
Sounds fine to leave as is but was also wondering about if we create and store a `RuntimeTransaction<SanitizedVersionedTransaction>` inside `ImmutableDeserializedPacket::new` and then add a new `RuntimeTransaction<SanitizedTransaction>::try_from_ref` function which takes a `&RuntimeTransaction::<SanitizedVersionedTransaction>` and does the clone internally?

Reply:
This brings back some concern we had about adding `Clone` to the `TransactionMeta` which means we must be careful about the `Meta` size. Adding `Clone` now may it difficult for us to remove in the future.

See above comment, the IDP is short-lived and will be removed soon.